samurai: new package

sbstndb opened the pull request:
-	Add samurai : an HPC library of mesh and physics

<!--  
Remember that `spackbot` can help with your PR in multiple ways:
- `@spackbot help` shows all the commands that are currently available
- `@spackbot fix style` tries to push a commit to fix style issues in this PR
- `@spackbot re-run pipeline` runs the pipelines again, if you have write access to the repository 
-->

sbstndb committed:
samurai: new package

	-	Add samurai : an HPC library of mesh and physics

alecbcs commented on var/spack/repos/builtin/packages/samurai/package.py:
```diff
@@ -0,0 +1,63 @@
+    """Intervals coupled with algebra of set to handle adaptive
+    mesh refinement and operators on it"""
+
+    homepage = "https://github.com/hpc-maths/samurai"
+    url = "https://github.com/hpc-maths/samurai.git"
```
```suggestion
    homepage = "https://github.com/hpc-maths/samurai"
    git = "https://github.com/hpc-maths/samurai.git"
```

We should set the `git` property when pulling a package from a git repository instead of setting a url. 
[See the docs here for more detail.](https://spack.readthedocs.io/en/latest/packaging_guide.html#fetching-from-code-repositories)

alecbcs commented on var/spack/repos/builtin/packages/samurai/package.py:
```diff
@@ -0,0 +1,63 @@
+    homepage = "https://github.com/hpc-maths/samurai"
+    url = "https://github.com/hpc-maths/samurai.git"
+
+    version("master", git=url)
```
```suggestion
    version("master", branch="master")
```

Once you set the `git` property in the package class it doesn't need to be reiterated here. However, we should specify the branch to use for the matching version like so.

alecbcs commented on var/spack/repos/builtin/packages/samurai/package.py:
```diff
@@ -0,0 +1,63 @@
+
+    version("master", git=url)
+
+    depends_on("xtl@0.7.4")
+    # depends_on("xsimd@11.0.0")
+    depends_on("xtensor@0.24.1 ~tbb")
+    depends_on("highfive~mpi", when="~mpi")
+    depends_on("highfive+mpi", when="+mpi")
+    depends_on("pugixml")
+    depends_on("fmt")
+    depends_on("nlohmann-json")
+    depends_on("cli11")
+
+    # option in future release
+    depends_on("cxxopts")
+    depends_on("cgal")
+    depends_on("petsc~mpi", when="~mpi")
+    depends_on("petsc+mpi", when="+mpi")
+    depends_on("boost+serialization+mpi", when="+mpi")
+
+    variant("mpi", default=False, description="Enable MPI support")
+    variant("openmp", default=False, description="Enable OpenMP support")
+    # variant("demos", default=False, description="Build Demos")
+    # variant("benchmarks", default=False,description="Build benchmarks")
+    variant("tests", default=False, description="Build tests")
+    variant("check_nan", default=False, description="Check for Nan in computations")
```
```suggestion
    variant("mpi", default=False, description="Enable MPI support")
    variant("openmp", default=False, description="Enable OpenMP support")
    # variant("demos", default=False, description="Build Demos")
    # variant("benchmarks", default=False,description="Build benchmarks")
    variant("tests", default=False, description="Build tests")
    variant("check_nan", default=False, description="Check for Nan in computations")
    
    depends_on("xtl@0.7.4")
    # depends_on("xsimd@11.0.0")
    depends_on("xtensor@0.24.1 ~tbb")
    depends_on("highfive~mpi", when="~mpi")
    depends_on("highfive+mpi", when="+mpi")
    depends_on("pugixml")
    depends_on("fmt")
    depends_on("nlohmann-json")
    depends_on("cli11")

    # option in future release
    depends_on("cxxopts")
    depends_on("cgal")
    depends_on("petsc~mpi", when="~mpi")
    depends_on("petsc+mpi", when="+mpi")
    depends_on("boost+serialization+mpi", when="+mpi")
```

Variants should be placed before `depends_on` and after `version` directives.

Should these commented variants be pulled out or uncommented? Similar concerns for the dependencies for future releases. Should those be included in this version or withheld for a future release?

sbstndb replied:
Thank you for the recommendation. The benchmark and demos variants have been disabled while they are being set up and will be added in the future, along with the xsimd dependency. I also need to find a way to copy the benchmark and demos binaries to the package location before enabling them

alecbcs commented on var/spack/repos/builtin/packages/samurai/package.py:
```diff
@@ -0,0 +1,63 @@
+        options.append(self.define_from_variant("SAMURAI_CHECK_NAN", "check_nan"))
+
+        # MPI support
+        if "+mpi" in spec:
```
```suggestion
        if spec.satisfies("+mpi"):
```

It's better style to use `spec.satisfies()` instead of `x in spec`. See the [docs here](https://spack.readthedocs.io/en/latest/packaging_guide.html#testing-spec-constraints) for more info.

alecbcs commented on var/spack/repos/builtin/packages/samurai/package.py:
```diff
@@ -0,0 +1,63 @@
+            options.append(self.define("HDF5_IS_PARALLEL", True))
+
+        # OpenMP support
+        if "+openmp" in spec:
```
```suggestion
        if spec.satisfies("+openmp"):
```

alecbcs reviewed (changes requested):
@sbstndb welcome to the Spack Community! This PR is looking really good. Just a couple of nitpicks and styling changes and then I think we'll be good to go.

sbstndb committed:
Update var/spack/repos/builtin/packages/samurai/package.py

Co-authored-by: Alec Scott <hi@alecbcs.com>

sbstndb committed:
Update var/spack/repos/builtin/packages/samurai/package.py

Co-authored-by: Alec Scott <hi@alecbcs.com>

sbstndb committed:
Update var/spack/repos/builtin/packages/samurai/package.py

Co-authored-by: Alec Scott <hi@alecbcs.com>

sbstndb committed:
Update var/spack/repos/builtin/packages/samurai/package.py

Co-authored-by: Alec Scott <hi@alecbcs.com>

sbstndb committed:
Update var/spack/repos/builtin/packages/samurai/package.py

Co-authored-by: Alec Scott <hi@alecbcs.com>

sbstndb committed:
Remove Whitespace

	-	Remove whitespace for spack style check

alecbcs commented on var/spack/repos/builtin/packages/samurai/package.py:
```diff
@@ -0,0 +1,63 @@
+    homepage = "https://github.com/hpc-maths/samurai"
+    git = "https://github.com/hpc-maths/samurai.git"
+
+    version("master", branch="master")
```
```suggestion
    version("master", branch="master")
    version("0.21.1", tag="v0.21.1", commit="29798bd9592e5b80674c49080a242378a61878c3")
```

Let's add a stable version here too, just for folks who might want to use but not actively develop the package.

alecbcs commented on var/spack/repos/builtin/packages/samurai/package.py:
```diff
@@ -0,0 +1,63 @@
+    depends_on("nlohmann-json")
+    depends_on("cli11")
+
+    # option in future release
```
What does this option mean? Are all of these dependencies optional? Or just `cxxopts`?

sbstndb committed:
Update var/spack/repos/builtin/packages/samurai/package.py

Co-authored-by: Alec Scott <hi@alecbcs.com>

sbstndb committed:
Add tags

	-	Add tags for the last versions of samurai
	-	All tags are tested and worked properly
	-	Add maintainers ("gouarin" - the samurai project lead and "sbstndb" - me, working on samurai)
	-	Add licence

alecbcs commented:
@gouarin please confirm you'd like to be made maintainer of this package with @sbstndb.

alecbcs reviewed (approved):
Will wait for confirmation from @gouarin. Otherwise this PR looks good to me.

gouarin commented:
I confirm! Thanks

Le dim. 23 févr. 2025, 19:56, Alec Scott ***@***.***> a
écrit :

> ***@***.**** approved this pull request.
>
> Will wait for confirmation from @gouarin <https://github.com/gouarin>.
> Otherwise this PR looks good to me.
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/spack/spack/pull/49144#pullrequestreview-2635716361>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/ABZJUFL4IDVLIEOZYHKC5QT2RIKXLAVCNFSM6AAAAABXTS677CVHI2DSMVQWIX3LMV43YUDVNRWFEZLROVSXG5CSMV3GSZLXHMZDMMZVG4YTMMZWGE>
> .
> You are receiving this because you were mentioned.Message ID:
> ***@***.***>
>

gouarin commented:
Do we need to do anything to complete the process?

alecbcs commented:
Nope! Sorry for the delay, this fell off of my radar.
